### DISCUSSION TITLE
updating features in a layer and adding to that layer while running

### DISCUSSION THREAD

**Discussion by gigfy2** [Mapsui/Mapsui#2344] [Q&A] [answered]
I've played with the samples in the mapsui repo and am now starting to try stuff in my own project using mapsui.  I can place svg/bitmaps on a map, but I'm not seeing how to manipulate that image object and update it's location, rotation, color, after it's placed.

I create a memory layer with Features and Style.  The Feature has where to place the image.  The Style contains what image to use, sizing and color.

So my questions are, 

1. How do I access these features in this layer so I can control them dynamically by updating their locations, rotation, color?

2. How can I dynamically create NEW features.  So, if I create 3 features when I first create the layer, how can I later add more, or remove existing ones?

I'm fairly new to mapsui, so this may have a simple answer, but I haven't figured it out yet

**Comment by charlenni** [comment]
If you could create the features for a MemoryLayer you should also capable of accessing them. Save them or look into the Features of this layer.

You could add more in the same way you added the first. I don't know how you did this, so it is very hard to suggest a way.

**Reply by gigfy2** [reply]
I'm still not able to create features and then after creation update them.  Here's the code I have

in the constructor I have

`mapControl.Map.Layers.Add(CreateSvgLayer());`

that calls

        private MemoryLayer CreateSvgLayer()
        {
            return new MemoryLayer
            {
                Name = "Points",
                IsMapInfoLayer = true,
                Features = GetPointFeatures(),
                Style = CreateBitmapStyle()
            };
        }

        private static IEnumerable<IFeature> GetPointFeatures()
        {
            var lon = 0;
            var lat = 0;

            List<IFeature> features = new List<IFeature>();

            var feature1 = new PointFeature(SphericalMercator.FromLonLat(lon, lat).ToMPoint());
            feature1["feature1key"] = "feature1";
            features.Add(feature1);

            var feature2 = new PointFeature(SphericalMercator.FromLonLat(lon +1, lat +1).ToMPoint());
            feature2["feature2key"] = "feature2";
            features.Add(feature2);

            var feature3 = new PointFeature(SphericalMercator.FromLonLat(lon-1, lat-1).ToMPoint());
            feature3["feature3key"] = "feature3";
            features.Add(feature3);

            return features;
        }

        private SymbolStyle CreateBitmapStyle()
        {
            var bitmapId = typeof(SvgSample).LoadBitmapId(@"Images.object.svg");

            return new SymbolStyle
            {
                BitmapId = bitmapId,
                SymbolOffset = new RelativeOffset(0.0, 0.5),
                SymbolScale = .1,
                BlendModeColor = Color.Red ?? Color.FromArgb(255, 57, 115, 199)
            };
        }

I've tried a bunch of stuff and nothing allows me to access those features I created.  If I watch mapControl.Map.Layers[1] (this svg layer is layer 1) while debugging in the dropdown I can see Features, and if I drop that down, it shows all 3.  But, I cannot figure out how to access the features in code so I can modify them during runtime.

I'm also not certain whether I should be using Layer, MemoryLayer, or some other type of layer.  I'm trying to do something that seems simple - add an svg to the map and then move it around - but, I can't seem to do it.  Any help is appreciated.

**Reply by gigfy2** [reply]
I finally got something to work to move an object 

            var e = mapControl.Map.Layers[1].Extent;

            var feature = mapControl.Map.Layers[1].GetFeatures(e, 1);

            var lon = 0;
            var lat = 0;
            var newPosition = new PointFeature(SphericalMercator.FromLonLat(lon - 60, lat).ToMPoint());

            foreach (var _feature in feature)
            {
                index++;

                if (index == 1)
                {
                    ((PointFeature)_feature).Point.X = newPosition.Point.X;
                    ((PointFeature)_feature).Point.Y = newPosition.Point.Y;
                }
            }
            
It works, but feels a little hacky.  Is there a better way to do this, or is this it?

**Comment by charlenni** [comment]
((List<Feature>)mapControl.Map.Layers[1].Features)[0]
    ((List<Feature>)mapControl.Map.Layers[1].Features)[1]
    ((List<Feature>)mapControl.Map.Layers[1].Features)[2]

are your 3 features. Or you give your features the same attribute by

    feature1["ID"] = "feature1";
    feature2["ID"] = "feature2";
    feature3["ID"] = "feature3";

and access it later by

    var feature = mapControl.Map.Layers[1].Features.Where((f) => f["ID"] == "feature2");

A MemoryLayer is in your case the right layer type.

I hope, that this is all correct. Didn't tested it.

**Reply by gigfy2** [reply]
In both of those cases, there's not a Feature available.  What you have makes logical sense, and I've tried to access it in similar ways, but  

    var feature = mapControl.Map.Layers[1].

there's a GetFeatures(), which is the hacky way I got something to work (posted that above), but no Features

**Comment by charlenni** [comment, marked as answer]
Correct. `mapControl.Map.Layers[1]` is of type `ILayer`. So you have to do

    var feature = ((List<Feature>)((MemoryLayer)mapControl.Map.Layers[1]).Features)[0]

Does this work?

**Reply by gigfy2** [reply]
doing this

    var feature = ((List<IFeature>)((MemoryLayer)mapControl.Map.Layers[1]).Features)[0];

    var newPosition = new PointFeature(SphericalMercator.FromLonLat(lon, lat+1).ToMPoint());

    ((PointFeature)feature).Point.X = newPosition.Point.X;
    ((PointFeature)feature).Point.Y = newPosition.Point.Y;

I can update the location of the feature.  So, that's awesome.  

But, when I try to access a feature by the key, I apparently have the syntax wrong.  I've tried several different ways, but nothing returns just the one feature.  One way I tried was

    var ftr2 = ((List<IFeature>)((MemoryLayer)mapControl.Map.Layers[1]).Features).Where(_ => _["featurekey"] == "featurekey2");

when the features were created, this is the one I'm trying to get

    var feature2 = new PointFeature(SphericalMercator.FromLonLat(lon +1, lat +1).ToMPoint());
    feature2["featurekey"] = "feature2";
    features.Add(feature2);
  
is the syntax wrong?

**Comment by charlenni** [comment]
Perhaps because you set 

    feature2["featurekey"] = "feature2";

and search for

    feature2["featurekey"] = "featurekey2"

    feature2["featurekey"] = "feature2";

**Reply by gigfy2** [reply]
ah, I was doing where on a list instead of firstordefault.  This worked

    var ftr5 = ((List<IFeature>)((MemoryLayer)mapControl.Map.Layers[1]).Features).FirstOrDefault(_ => _["featurekey"] == "feature2");
    
but yes, you were right I was searching for the wrong string, so FirstOrDefault instead of Where, and search for the correct string lol

Thank you for all your help charlenni.  I've been stuck here for a while and now I can move on.  I really appreciate the help
